### PR TITLE
Support for RE3 exploration algorithm

### PR DESCRIPTION
**Pull request by n30111**
## Why are these changes needed?

This PR adds support for RE3 (Random Encoders for Efficient Exploration). RE3 is a simple and efficient algorithm for off-policy RL algorithms. This can also be used with on-policy RL algorithms. 

This is the implementation of  **State entropy maximization with random encoders for efficient exploration**. Seo, Chen, Shin, Lee, Abbeel, & Lee, (2021). arXiv preprint arXiv:2102.09430.
## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/agents/trainer.py
Comment (gjoliver) [inline rllib/agents/trainer.py:829]: I think you can just use policy.global_timestep and there is no need to track this yourself?
another way is to Add a MixIn for this Beta schedule, much like LearningRateSchedule and EntropyCoeffSchedule. you can then just use the auto-decayed beta value here.

File: rllib/agents/trainer.py
Comment (gjoliver) [inline rllib/agents/trainer.py:804]: this feels a bit weird.
I guess to calculate the knn distance, you want to use a batch that's randomly sampled from the ReplayBuffer? just so you can measure how different current sample batch is from all the things you have been seeing (RB)?
as written, this knn thing is calculated from a single batch. so naturally, you are gonna get a lot of similar samples from consecutive steps?

File: rllib/agents/trainer.py
Comment (n30111) [reply rllib/agents/trainer.py:778]: Yes, that would be cleaner. I will make the changes.

File: rllib/agents/trainer.py
Comment (n30111) [reply rllib/agents/trainer.py:829]: Here, we want to decay based on `training_iteration`. It seems `policy.global_timestep` depends on number of batch processed and not equal to `training_iteration`, just looking at source, not 100% sure.

File: rllib/agents/trainer.py
Comment (n30111) [reply rllib/agents/trainer.py:804]: Yes, KNN is computed for a single batch and we are assuming that RLLib has sampled that batch randomly from replay buffer.

Comment (n30111) [reply rllib/agents/trainer.py:804]: KNN distance and the randomness is limited to single batch only, not the full replay buffer.

File: rllib/policy/policy.py
Comment (n30111) [reply rllib/policy/policy.py:700]: updated

File: rllib/policy/sample_batch.py
Comment (n30111) [reply rllib/policy/sample_batch.py:44]: updated

**Comment by jbedorf** [comment]
@gjoliver  Could you take another look and indicate if anything else needs to be changed? Thanks!

**Comment by gjoliver** [comment]
sorry about the delay, I will take another detailed look.

File: rllib/examples/re3_exploration.py
Comment (gjoliver) [inline rllib/examples/re3_exploration.py:28]: sac.DEFAULT_CONFIG doesn't specify callbacks?
it seems like you can simply do:
```
config["callbacks"] = partial(RE3Callbacks, embeds_dim=128,
    beta_schedule="linear_decay", k_nn=50)
```
here.

File: rllib/agents/trainer.py
Comment (gjoliver) [inline rllib/agents/trainer.py:767]: can you undo this empty line diff too? thanks a lot.

File: rllib/utils/exploration/random_encoder.py
Comment (gjoliver) [inline rllib/utils/exploration/random_encoder.py:60]: hmm sorry for the stupid question, but I don't quite get the formula for the moving mean here. as written, self.mean basically is:

self.mean + batch_mean - self.mean + batch_count / tot_count
which is:
batch_mean + batch_count / tot_count

why are we taking batch_mean plus a small ratio as the moving mean?

File: rllib/utils/exploration/random_encoder.py
Comment (gjoliver) [inline rllib/utils/exploration/random_encoder.py:92]: add whitespaces around ** ?

File: rllib/utils/exploration/random_encoder.py
Comment (gjoliver) [inline rllib/utils/exploration/random_encoder.py:166]: on_train_result() doesn't necessarily fire after each iteration though ...
I think the best way to do this is to pass trainer._iteration as a parameter to on_learn_on_batch() call.
that way, you don't even need to track it yourself.

maybe you can add a TODO here for me. I am happy to refactor and clean this up once this PR is merged. it feels like a bigger change that we should help you guys.

File: rllib/utils/exploration/random_encoder.py
Comment (gjoliver) [inline rllib/utils/exploration/random_encoder.py:162]: maybe you should call super() first, before your entropy logics here.
like for example, in case some other callbacks have their own opinions about REWARDS and ADVANTAGES, you will get the final say.

File: rllib/utils/exploration/random_encoder.py
Comment (gjoliver) [inline rllib/utils/exploration/random_encoder.py:184]: should we mention here that entropy is only calculated per batch, and does not take the distribution of the entire replay buffer into consideration?
basically one of your responses, we just put it here.

File: rllib/utils/exploration/random_encoder.py
Comment (gjoliver) [inline rllib/utils/exploration/random_encoder.py:204]: update the comment? this implementation doesn't support torch.

File: rllib/utils/exploration/random_encoder.py
Comment (gjoliver) [inline rllib/utils/exploration/random_encoder.py:231]: actually can I ask why not?
if it's just a additional labor thing, maybe we can add a TODO.
otherwise, let's explain the choice a little bit.
thanks a lot.

File: rllib/utils/exploration/tests/test_random_encoder.py
Comment (gjoliver) [inline rllib/utils/exploration/tests/test_random_encoder.py:66]: move import to top of file?

File: rllib/utils/exploration/random_encoder.py
Comment (gjoliver) [inline rllib/utils/exploration/random_encoder.py:271]: nit picking.
it's not really "inside" ModelV2?
more like "Creates ModelV2 embedding module / layers"?

**Review by gjoliver — COMMENTED** [review]
this looks very exciting. thanks for your awesome work.

I have gone through the PR carefully, and had a bunch of minor / detailed comments. nothing significant actually, except maybe for a TODO for us :)

The biggest blocker right now is to rebase to master, and make sure all the tests pass. there are too many failed CI tests right now.
This PR doesn't introduce any changes to existing codebase, so I don't expect anything tricky. Probably just need to retry by doing:

fetch upstream in your repo
git checkout master; git pull --rebase
git checkout re3;  git rebase master
then git push to re-trigger the tests.

Also, please make sure you run ci/travis/format.sh so Lint test will pass too.

Thanks again. Let me know if you need any help with the tests actually.
Happy to help on those logistic things.

File: rllib/utils/exploration/random_encoder.py
Comment (n30111) [reply rllib/utils/exploration/random_encoder.py:166]: I see. I was not aware of a best way to access trainer._iteration inside `on_learn_on_batch` call. It would be great if refactor it as required.

File: rllib/utils/exploration/random_encoder.py
Comment (n30111) [reply rllib/utils/exploration/random_encoder.py:231]: I have not used pytorch yet, basically its a labor thing. We can add a TODO.

File: rllib/utils/exploration/random_encoder.py
Comment (n30111) [reply rllib/utils/exploration/random_encoder.py:60]: Yes, this construct is adopted author's reference code. I have also experimented using the correct moving mean (`(self.mean * self.count + batch_mean * batch_count) / tot_count`) version, results are almost similar.

File: rllib/examples/re3_exploration.py
Comment (n30111) [reply rllib/examples/re3_exploration.py:28]: This was used for demonstration purpose, in case user wants to know how it works when callbacks is provided in configs.

File: rllib/utils/exploration/random_encoder.py
Comment (n30111) [reply rllib/utils/exploration/random_encoder.py:96]: `ci/travis/format.sh` script removes the whitespaces automatically.

**Comment by n30111** [comment]
@gjoliver I have updated the code with requested changes, please have a look. It would be great if you can help with the failing tests. Not sure why these tests are failing. Thank you.

**Comment by gjoliver** [comment]
I can totally help. can you give me edit permission to this repo?
or if you can help pull down the latest master, that will be awesome.
I will push some minor changes to your branch.

thanks.

**Comment by jbedorf** [comment]
> I can totally help. can you give me edit permission to this repo? or if you can help pull down the latest master, that will be awesome. I will push some minor changes to your branch.
> 
> thanks.

I've added you to the repository. Let me know if it does not work.

File: rllib/utils/exploration/random_encoder.py
Comment (gjoliver) [reply rllib/utils/exploration/random_encoder.py:64]: ah ok. it's not uncommon to see questionable reference implementation.
we can keep it as is for now.

**Comment by gjoliver** [comment]
Tests are now all green!

**Review by gjoliver — APPROVED** [review]

File: rllib/agents/callbacks.py
Comment (sven1977) [inline rllib/agents/callbacks.py:390]: Could we move this into `utils/exploration/random_encoder.py`?

File: rllib/utils/exploration/tests/test_random_encoder.py
Comment (sven1977) [inline rllib/utils/exploration/tests/test_random_encoder.py:1]: This is great! Could we actually activate this test? By adding it to `rllib/BUILD` to the other `utils/exploration` tests?
This will make sure this feature never breaks in the future.

File: rllib/utils/exploration/random_encoder.py
Comment (sven1977) [reply rllib/utils/exploration/random_encoder.py:185]:
```suggestion
            raise ValueError("This RE3 implementation does not support PyTorch yet.")
```

**Review by sven1977 — COMMENTED** [review]
Hey @n30111 , thanks for this really great PR! Just 2-3 smaller nits, like activating the test case, etc...
In a follow up PR, could you also add this exploration algo to our algo docs? There is a special section in there describing the different exploration strategies.

File: rllib/agents/callbacks.py
Comment (n30111) [reply rllib/agents/callbacks.py:390]: It was moved out from here to avoid circular dependency in this commit https://github.com/ray-project/ray/pull/19551/commits/89169ccb63b0a780e3eb6231a38907244cfc1278

**Comment by n30111** [comment]
> Hey @n30111 , thanks for this really great PR! Just 2-3 smaller nits, like activating the test case, etc...
> In a follow up PR, could you also add this exploration algo to our algo docs? There is a special section in there describing the different exploration strategies.

Sure, I will update the docs in a follow up PR. Thanks.

File: rllib/agents/callbacks.py
Comment (gjoliver) [reply rllib/agents/callbacks.py:390]: yeah, this is the best I could do.
I would love to keep everything in random_encoder.py too, but you know :)

**Comment by n30111** [comment]
@gjoliver Can you please help again on the tests part, not sure why these tests are failing. Thanks.

**Comment by gjoliver** [comment]
> @gjoliver Can you please help again on the tests part, not sure why these tests are failing. Thanks.

can you update your test? the error message looks like:



from ray.rllib.utils.exploration.random_encoder import RE3UpdateCallbacks
--
  | ImportError: cannot import name 'RE3UpdateCallbacks' from 'ray.rllib.utils.exploration.random_encoder'

**Comment by n30111** [comment]
@gjoliver I was getting error due to the changes in this commit https://github.com/ray-project/ray/pull/19551/commits/491ef895c49d06dbdb05a728f1b3b556929ef4f9 . Do you think we need to revert this or is there any other solution?

**Review by sven1977 — APPROVED** [review]
Awesome, thanks! :)

File: rllib/policy/dynamic_tf_policy.py
Comment (sven1977) [inline rllib/policy/dynamic_tf_policy.py:683]: No need to do this rn, but in a follow up PR, we should also test-run the callbacks, so that the SampleBatch access detector captures this field and automatically adds it to the view-requirements (instead of always doing so).

File: rllib/BUILD
Comment (sven1977) [inline rllib/BUILD:1269]:
```suggestion
    size = "medium",
```

**Comment by sven1977** [comment]
Merging this now as all tests look good.
If you would like to add more changes, just do a follow-up PR.
Thanks again so much @n30111 . Really nice PR!

**Comment by n30111** [comment]
> Merging this now as all tests look good.
> If you would like to add more changes, just do a follow-up PR.
> Thanks again so much @n30111 . Really nice PR!

Thanks, great.

**Pull request merged**
